### PR TITLE
JENKINS-38609: Shared Library non ROOT support

### PR DESCRIPTION
**Pull request by heyleke**
This is a new implementation for JENKINS-38609, which makes it possible to load libraries from a specified relative path. Only touching files mentioned in review of #84

### CODE REVIEW

**Review by dwnusbaum — CHANGES_REQUESTED** [review]
Thanks for the PR! I added some minor comments about refactoring the setters, and a more important comment about making sure this PR doesn't introduce security issues, which needs to be addressed.

File: src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java [hunk 1]
```diff
@@ -157,14 +174,25 @@ static void doRetrieve(String name, boolean changelog, @Nonnull SCM scm, FilePat
...
+            LOGGER.info("libBase: "+libBase);
+
+            if (lease.path.child(libBase).child("src/test").exists()) {
                 listener.getLogger().println("Excluding src/test/ from checkout of " + scm.getKey() + " so that shared library test code cannot be accessed by Pipelines.");
                 listener.getLogger().println("To remove this log message, move the test code outside of src/. To restore the previous behavior that allowed access to files in src/test/, pass -D" + SCMSourceRetriever.class.getName() + ".INCLUDE_SRC_TEST_IN_LIBRARIES=true to the java command used to start Jenkins.");
             }
-            lease.path.copyRecursiveTo("src/**/*.groovy,vars/*.groovy,vars/*.txt,resources/", excludes, target);
+            // Cannot add WorkspaceActionImpl to private CpsFlowExecution.flowStartNodeActions; do we care?
+            // Copy sources with relevant files from the checkout:
+            lease.path.child(libBase).copyRecursiveTo("src/**/*.groovy,vars/*.groovy,vars/*.txt,resources/", excludes, target);
```
[the remark this answers is not in the document]
Comment (heyleke) [reply src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java:191]: I will check this 'offline', it is indeed not 'secure'

[the remark this answers is not in the document]
Comment (heyleke) [reply src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java:199]: taken care of in new commit (I squashed everything again)

**Comment by heyleke** [comment]
ping @dwnusbaum can you please re-review after changes done?

**Comment by dwnusbaum** [comment]
@heyleke I filed https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/101 based on this PR with various minor changes (the commit is still attributed to you), if you can review that PR it would be appreciated.

**Comment by mahmoud-ashi** [comment]
Hey all. Can we move this forward? It has been pending since some time now :). This change would really be nice for my use case. Thanks.

**Comment by jglick** [comment]
@car-roll since #101 was merged, should this not be closed?

**Pull request closed without merging**